### PR TITLE
Upgrade Azure client generation

### PR DESCRIPTION
**Pull request by hetalsonavane**
@dubuc @embik  Please review this and Let me know the required changes to be done

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+go 1.14
+
+require (
+	github.com/Azure/azure-sdk-for-go v42.3.0+incompatible
+	github.com/Azure/go-autorest/autorest v0.10.2
+	github.com/Azure/go-autorest/autorest/azure/auth v0.4.2
+	github.com/Azure/go-autorest/autorest/to v0.3.0 // indirect
+	github.com/Azure/go-autorest/autorest/validation v0.2.0 // indirect
+	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
+	github.com/hetalsonavane/azure-request-limitometer v0.0.0-20200522130424-944659dcfad6
```
Comment (embik) [inline go.mod]: This makes the code depend on your fork as a dependency, I think we don't want that 😄

File: internal/config/config.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
+package config
```
Comment (embik) [inline internal/config/config.go:1]: This was copied over from the Azure-Samples repository, so I think we need to add the license as a header to the files in question to ensure we're not violating a license.

File: go.mod (hunk 1 above)
Comment (embik) [reply go.mod]: It also stops me from building the project after checking it out at the moment:

```
go build .
go: downloading github.com/spf13/pflag v1.0.5
go: downloading github.com/Azure/go-autorest/autorest v0.10.2
go: downloading github.com/hetalsonavane/azure-request-limitometer v0.0.0-20200522130424-944659dcfad6
go: downloading github.com/Azure/go-autorest/autorest/adal v0.8.2
go: downloading github.com/Azure/go-autorest/logger v0.1.0
go: downloading github.com/Azure/go-autorest/tracing v0.5.0
go: finding module for package github.com/hetalsonavane/azure-request-limitometer/internal/config
go: downloading github.com/Azure/go-autorest/autorest/azure/auth v0.4.2
go: downloading github.com/Azure/azure-sdk-for-go v42.3.0+incompatible
go: downloading github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
go: downloading github.com/prometheus/client_golang v1.6.0
go: downloading github.com/influxdata/influxdb v1.8.0
go: downloading github.com/Azure/go-autorest/autorest/date v0.2.0
go: downloading github.com/dgrijalva/jwt-go v3.2.0+incompatible
go: downloading github.com/dimchansky/utfbom v1.1.0
go: downloading golang.org/x/crypto v0.0.0-20191206172530-e9b2fee46413
go: downloading github.com/Azure/go-autorest/autorest/azure/cli v0.3.1
go: downloading github.com/prometheus/common v0.9.1
go: downloading github.com/prometheus/procfs v0.0.11
go: downloading github.com/golang/protobuf v1.4.0
go: downloading github.com/beorn7/perks v1.0.1
go: downloading google.golang.org/protobuf v1.21.0
go: downloading github.com/prometheus/client_model v0.2.0
go: downloading golang.org/x/sys v0.0.0-20200420163511-1957bb5e6d1f
go: downloading github.com/cespare/xxhash v1.1.0
go: downloading github.com/mitchellh/go-homedir v1.1.0
go: downloading github.com/matttproud/golang_protobuf_extensions v1.0.1
go: downloading github.com/cespare/xxhash/v2 v2.1.1
go: downloading github.com/hetalsonavane/azure-request-limitometer v0.0.0-20200528141315-1b409f067172
go: downloading github.com/Azure/go-autorest/autorest/to v0.3.0
go: downloading github.com/Azure/go-autorest/autorest/validation v0.2.0
go: found github.com/hetalsonavane/azure-request-limitometer/internal/config in github.com/hetalsonavane/azure-request-limitometer v0.0.0-20200528141315-1b409f067172
go: azure-request-limitometer/cmd/limitometer imports
	github.com/hetalsonavane/azure-request-limitometer/internal/config: github.com/hetalsonavane/azure-request-limitometer@v0.0.0-20200528141315-1b409f067172: parsing go.mod:
	module declares its path as: azure-request-limitometer
	        but was required as: github.com/hetalsonavane/azure-request-limitometer
```

**Comment by hetalsonavane** [comment]
@dubuc  @embik  Please review updated changes. Change in GetAllVm and GetAllNic method

File: go.mod [hunk 3]
```diff
@@ -1,18 +1,15 @@
-module github.com/cerence/azure-request-limitometer
+module azure-request-limitometer
```
Comment (embik) [inline go.mod]: Please ensure the module is named github.com/cerence/azure-request-limitometer.

File: go.mod [hunk 4]
```diff
@@ -1,18 +1,15 @@
...
-	github.com/Azure/go-autorest/autorest/to v0.3.0 // indirect
-	github.com/Azure/go-autorest/autorest/validation v0.2.0 // indirect
 	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
-	github.com/influxdata/influxdb v1.7.0
-	github.com/influxdata/platform v0.0.0-20181107003602-9b529771ebb3 // indirect
-	github.com/prometheus/client_golang v0.9.3
-	github.com/prometheus/common v0.4.1 // indirect
-	github.com/prometheus/procfs v0.0.2 // indirect
-	github.com/spf13/pflag v1.0.3
+	github.com/hetalsonavane/azure-request-limitometer v0.0.0-20200624102419-6b2358ab5938 // indirect
```
Comment (embik) [inline go.mod]: This is caused by module naming issues. Fix the module name and run "go mod tidy" afterwards to clean this up, please.

File: pkg/common/client.go [hunk 5]
```diff
@@ -209,6 +152,57 @@ func getLastSegment(ID string) (string, error) {
...
+// PutVM returns the Virtual Machine object
+func (az AzureClient) PutVM(nodename string) (res autorest.Response) {
+	ctx, cancel := context.WithTimeout(context.Background(), 6000*time.Second)
+	defer cancel()
+	fmt.Println("PutVm start")
+	node, err := az.GetVM(ctx, nodename)
+	if err != nil {
+		log.Panic(err)
+	}
+	fmt.Println("Putvm Middle")
```
Comment (embik) [inline pkg/common/client.go]: looks like left-over debugging code to me.

File: pkg/common/client.go [hunk 6]
```diff
@@ -209,6 +152,57 @@ func getLastSegment(ID string) (string, error) {
...
+		log.Panicf("failed to get all VMs: %v", err)
+	}
+	return
+}
+
+// PutVM returns the Virtual Machine object
+func (az AzureClient) PutVM(nodename string) (res autorest.Response) {
+	ctx, cancel := context.WithTimeout(context.Background(), 6000*time.Second)
+	defer cancel()
+	fmt.Println("PutVm start")
```
Comment (embik) [inline pkg/common/client.go]: looks like left-over debugging code to me.

File: pkg/common/client.go [hunk 7]
```diff
@@ -209,6 +152,57 @@ func getLastSegment(ID string) (string, error) {
...
+	node, err := az.GetVM(ctx, nodename)
+	if err != nil {
+		log.Panic(err)
+	}
+	fmt.Println("Putvm Middle")
+	req, err := az.VirtualMachinesClient.CreateOrUpdatePreparer(ctx,Client.config.ResourceGroup, nodename, node)
+	if err != nil {
+		log.Panic(err)
+	}
+	fmt.Println("Actual Putvm req",req)
```
Comment (embik) [inline pkg/common/client.go]: looks like left-over debugging code to me.

File: cmd/limitometer/main.go [hunk 8]
```diff
@@ -9,27 +9,29 @@ import (
...
 
-var config = common.Conf
-var azureClient = common.Client
-
 var (
 	nodename     = flag.String("node", "", "Valid node in the resource group to create compute queries. Environment Variable: NODE_NAME")
 	target       = flag.String("output", "pushgateway", "Target output for the limitometer, supported values are: [influxdb|pushgateway]")
 	mode         = flag.String("mode", "oneshot", "Operational mode for limitometer, supported values are: [oneshot|service]")
 	pollInterval = flag.Int("poll-interval", 60, "Only for 'service' mode: Poll interval for refreshing metrics in seconds")
+	configBy     = flag.String("configuration", "loadconfig", "To decide from where to load config, supported values are: [loadconfig|environment]")
```
Comment (embik) [inline cmd/limitometer/main.go]: Can we rename the mode to "metadata" instead?

File: cmd/limitometer/main.go [hunk 9]
```diff
@@ -75,8 +77,20 @@ func main() {
 		*nodename = env
 	}
 
-	log.Printf("Starting limitometer with %s as target VM", *nodename)
+	if strings.ToLower(*configBy) == "loadconfig" {
+		common.Conf = common.LoadConfig()
```
Comment (embik) [inline cmd/limitometer/main.go]: Why is this required? The `NewClient` function is calling `LoadConfig` as well, so this looks like duplication. Could you explain?

**Review by embik — CHANGES_REQUESTED** [review]

File: pkg/common/client.go (hunk 5 above)
Comment (hetalsonavane) [reply pkg/common/client.go]: Solved

File: pkg/common/client.go (hunk 6 above)
Comment (hetalsonavane) [reply pkg/common/client.go]: updated.

File: cmd/limitometer/main.go (hunk 9 above)
Comment (hetalsonavane) [reply cmd/limitometer/main.go]: Yes you are right !! I have removed it from main file as in NewClient it will initialize config object depending on metadata argument.

File: cmd/limitometer/main.go (hunk 8 above)
Comment (hetalsonavane) [reply cmd/limitometer/main.go]: I updated it to metadata.

File: go.mod (hunk 1 above)
Comment (hetalsonavane) [reply go.mod]: Updated to module github.com/cerence/azure-request-limitometer

File: go.mod (hunk 4 above)
Comment (hetalsonavane) [reply go.mod]: I have updated  go.mod

File: internal/config/config.go (hunk 2 above)
Comment (hetalsonavane) [reply internal/config/config.go:1]: Added comment please review and lemme know if you want any changes.

File: go.mod (hunk 3 above)
Comment (hetalsonavane) [reply go.mod]: Updated module github.com/cerence/azure-request-limitometer

File: cmd/limitometer/main.go [hunk 10]
```diff
@@ -9,27 +9,28 @@ import (
...
 
-var config = common.Conf
-var azureClient = common.Client
-
 var (
 	nodename     = flag.String("node", "", "Valid node in the resource group to create compute queries. Environment Variable: NODE_NAME")
 	target       = flag.String("output", "pushgateway", "Target output for the limitometer, supported values are: [influxdb|pushgateway]")
 	mode         = flag.String("mode", "oneshot", "Operational mode for limitometer, supported values are: [oneshot|service]")
 	pollInterval = flag.Int("poll-interval", 60, "Only for 'service' mode: Poll interval for refreshing metrics in seconds")
+	metadata     = flag.String("metadata-config", "loadconfig", "To decide from where to load config, supported values are: [loadconfig|environment]")
```
Comment (embik) [inline cmd/limitometer/main.go]: Sorry, I was trying to say that `--config=metadata` and `--config=environment` should be valid. Sorry for the confusion :)

File: cmd/limitometer/main.go [hunk 11]
```diff
@@ -74,9 +75,24 @@ func main() {
 	if exists {
 		*nodename = env
 	}
+	confval, exists := os.LookupEnv("METADATA")
```
Comment (embik) [inline cmd/limitometer/main.go]: Can we name this `LIMITOMETER_CONFIG` maybe?

File: internal/config/config.go (hunk 2 above)
Comment (embik) [reply internal/config/config.go:1]: I think I would prefer if it said "this file has been adapted from azure-sdk-for-go-samples, licensed under the MIT license" or something like that.

File: cmd/limitometer/main.go (hunk 10 above)
Comment (hetalsonavane) [reply cmd/limitometer/main.go]: Updated the variable as well as name.

File: cmd/limitometer/main.go (hunk 11 above)
Comment (hetalsonavane) [reply cmd/limitometer/main.go]: Updated value to  LIMITOMETER_CONFIG

File: internal/config/config.go (hunk 2 above)
Comment (hetalsonavane) [reply internal/config/config.go:1]: Updated the comments.

File: pkg/common/client.go [hunk 12]
```diff
@@ -19,172 +18,117 @@ import (
...
-	defer cancel()
-
-	vm, err := c.VirtualMachinesClient.Get(ctx, Conf.ResourceGroup, vmname, compute.InstanceView)
+//GetVMClient return vmClient
+func GetVMClient(configload Config) compute.VirtualMachinesClient {
+	vmClient := compute.NewVirtualMachinesClient(configload.SubscriptionID)
+	a, err := auth.NewAuthorizerFromEnvironment()
 	if err != nil {
-		log.Panicf("failed to get VM: %v", err)
+		glog.Fatalf("failed to create authorizer from environment: %s\n", err)
```
Comment (embik) [inline pkg/common/client.go]: Just realized this, why are we switching from log to glog here?

File: pkg/common/client.go [hunk 13]
```diff
@@ -19,172 +18,117 @@ import (
...
+	nicClient := network.NewInterfacesClient(configload.SubscriptionID)
+	a, err := auth.NewAuthorizerFromEnvironment()
 	if err != nil {
-		log.Panicf("failed to get all VMs: %v", err)
+		glog.Fatalf("failed to create authorizer from environment: %s\n", err)
 	}
-
-	return
+	nicClient.Authorizer = a
+	nicClient.AddToUserAgent("sdk-samples")
```
Comment (embik) [inline pkg/common/client.go]: Can we use "azure-request-limitometer" as user agent please?

File: pkg/common/client.go [hunk 14]
```diff
@@ -19,172 +18,117 @@ import (
...
-	res.Response = result
-
-	return
-}
-
-// GetNicFromVMName returns primary nic object based on vm name
-func (c AzureClient) GetNicFromVMName(vmName string) network.Interface {
-	return c.getNic(vmName, true)
+	lbClient.Authorizer = a
+	lbClient.AddToUserAgent("sdk-samples")
```
Comment (embik) [inline pkg/common/client.go]: same here, let's use another user agent.

**Review by embik — CHANGES_REQUESTED** [review]

File: pkg/common/client.go (hunk 12 above)
Comment (hetalsonavane) [reply pkg/common/client.go]: Not intentionally .

File: pkg/common/client.go (hunk 14 above)
Comment (hetalsonavane) [reply pkg/common/client.go]: updated with azure-request-limitometer-Lb

File: pkg/common/client.go (hunk 13 above)
Comment (hetalsonavane) [reply pkg/common/client.go]: Updated with azure-request-limitometer-Nic

File: pkg/common/client.go (hunk 12 above)
Comment (embik) [reply pkg/common/client.go]: Could we use log in that case to be consistent? :)

Comment (embik) [reply pkg/common/client.go]: There's a few other occasions where code was changed to glog I think.

Comment (hetalsonavane) [reply pkg/common/client.go]: Sure I can use .

**Review by embik — APPROVED** [review]
